processor/src/trace/range/mod.rs

```diff
@@ -0,0 +1,55 @@
+use super::{Felt, FieldElement};
+
+use crate::range::AuxColumnHint;
+
+pub use vm_core::range::{P0_COL_IDX, P1_COL_IDX, T_COL_IDX, V_COL_IDX};
+
+#[cfg(test)]
+mod tests;
+
+// HELPER FUNCTIONS
+// ================================================================================================
+
+/// Builds the execution trace of the range checker's `p0` auxiliary column used for multiset
+/// checks. The running product is built up in the 8-bit section of the table and reduced in the
+/// 16-bit section of the table so that the starting and ending value are both one.
+pub fn build_aux_col_p0<E: FieldElement<BaseField = Felt>>(
+    aux_column: &mut [E],
+    aux_column_hints: &[AuxColumnHint],
+    rand_elements: &[E],
+    t_col: &[Felt],
+    v_col: &[Felt],
+) {
+    aux_column[0] = E::ONE;
+
+    // Build the execution trace of the 8-bit running product.
+    for (row_idx, hint) in aux_column_hints
+        .iter()
+        .enumerate()
+        .take(aux_column.len() - 1)
+    {
+        let t: E = t_col[row_idx].into();
+        let v: E = v_col[row_idx].into();
+        let alpha = rand_elements[0];
+
+        if t == E::ZERO {
+            // This is the 8-bit section, where the running product must be built up.
+            // Define variable z as: z = f3​*(α+v)^4 + f2*(α+v)^2 ​+ f1*(α+v) ​+ f0​
+            let z = match hint {
+                AuxColumnHint::F0 => E::ONE,
+                AuxColumnHint::F1 => v + alpha,
+                AuxColumnHint::F2 => (v + alpha).square(),
+                AuxColumnHint::F3 => ((v + alpha).square()).square(),
+            };
+
+            aux_column[row_idx + 1] = aux_column[row_idx] * z;
+        } else if t == E::ONE {
+            // This is the 16-bit section, where the running product must be reduced.
+            let v_next = v_col[row_idx + 1].into();
+            aux_column[row_idx + 1] = aux_column[row_idx] / (alpha + v_next - v);
```

Comment:
For large traces, doing division at every row is going to be very expensive (I think cost of a division is roughly equal to the cost of 80 multiplications - though, it might be somewhat different in extension fields). A better approach could be:
* Compute `(alpha + v_next - v)` for each row and save the results in a vector.
* Run batch inversion on this vector. Instead of *n* inversions, this would require *3n* multiplications + 1 inversion.
* Multiply the aux_column values by respective `1 / (alpha + v_next - v)`.